proposals/NNNN-library-evolution.md

```diff
@@ -0,0 +1,374 @@
+# Library Evolution for Stable ABIs
+
+* Proposal: [SE-NNNN](NNNN-non-exhaustive-enums.md)
+* Authors: [Jordan Rose](https://github.com/jrose-apple)
+* Review Manager: TBD
+* Status: **Awaiting review**
+* Implementation: Implemented in Swift 5 for standard library use
+* Pre-review discussion: (TODO)
+
+<!--
+*During the review process, add the following fields as needed:*
+
+* Decision Notes: [Rationale](https://lists.swift.org/pipermail/swift-evolution/), [Additional Commentary](https://lists.swift.org/pipermail/swift-evolution/)
+* Bugs: [SR-NNNN](https://bugs.swift.org/browse/SR-NNNN), [SR-MMMM](https://bugs.swift.org/browse/SR-MMMM)
+* Previous Revision: [1](https://github.com/apple/swift-evolution/blob/...commit-ID.../proposals/NNNN-filename.md)
+* Previous Proposal: [SE-XXXX](XXXX-filename.md)
+-->
+
+## Introduction
+
+One of Swift's goals is to be a good language for libraries with binary compatibility concerns, such as those shipped as part of Apple's OSs. This includes giving library authors the flexibility to add to their public interface, and to change implementation details, without breaking binary compatibility. At the same time, it's important that library authors be able to opt out of this flexibility in favor of performance.
+
+This proposal introduces:
+
+- a "library evolution" build mode for libraries that are declaring ABI stability, which preserves the ability to make certain changes to types without breaking the library's ABI; and
+- an attribute for such libraries to opt out of this flexibility on a per-type basis, allowing certain compile-time optimizations.
+
+The mechanisms for this are already in place, and were used to stabilize the
+ABI of the standard library. This proposal makes them features for use by
+any 3rd-party library that wishes to declare itself ABI stable.
+
+**This build mode will have no impact on libraries built and distributed with an app**. Such libraries will receive the same optimization that they have in previous versions of Swift.
+
+_Note: this proposal will use the word "field" to mean "stored instance property"._
+
+## Motivation
+
+Starting in Swift 5, libraries are able to declare a stable ABI, allowing a  library binary to be replaced with a newer version without requiring client programs to be recompiled. 
```

Comment:
"Starting in Swift 5" might not really make sense now.

Reply:
"as of" I guess

Reply:
Ah, I meant that it's something we did for the standard library and overlays, but that isn't generally available.